### ISSUE TITLE
[Question]: NVME ? USB ?

### ISSUE THREAD

**Issue by toxic0berliner** [vdsm/virtual-dsm#780]
### Is your question not already answered in the FAQ?

- [X] I made sure the question is not listed in the [FAQ](https://github.com/vdsm/virtual-dsm/blob/master/readme.md).

### Is this a general question and not a technical issue?

- [X] I am sure my question is not about a technical issue.

### Question

Hi,

I was aiming at replacing my physical DS918+ with a VM runing vDSM, happy that if I get a new physical syno NAS in the future I can simply unplug my 20TB drive and replug it into a real syno if they manage to sell me a device with 10G networking at a reasonable price one day...

I have managed to get VDSM up and running, but my goal is to use it passing it the physical HDD drive entirely, works fine, but performance is way below my actual physical NAS.

Indeed, my physical NAS is limited by it's gigabit NIC, but it has NVME write cache allowing it to sustain gigabit for a the full file transfer (smb, nfs, ...), I'm sseing 118Mio/s, gigabit line speed. I was aiming at improving that with VDSM, and to an extent it does, as I'm seeing 240Mio/s at the beginning of the file transfer but very quickly it drops down to 83Mio/s which I believe is the hard limit for my single 20TB SATA HDD...

I managed to pass 2 NVME drives to my qemu vDSM VM (in PVE) by adding the following args: 
`-drive file=/mnt/nvme/vdsm-nvme-0.qcow2,if=none,id=NVME1 -device nvme,drive=NVME1,serial=nvme-1 -drive file=/mnt/nvme2/vdsm-nvme-1.qcow2,if=none,id=NVME2 -device nvme,drive=NVME2,serial=nvme-2`

But seems VDSM does not see it or decides not to use it, or simply does not include the software to make use of nvme read or write cache...

I understand this makes sens when running vdsm on a syno device as one would pass a volume to vdsm which itself might already be a HDD accelerated with nvme cache, but in my case, if I want vdsm to have good read/write performance I will need to pass it a virtual disk stored manages by my underlying PVE as zfs with caching and stuff... This would also mean that I will not be able to take this physical HDD out of my PVE host and simply plug it into a physical synology if I want to migrate easily in the future (or if my cheap hardware for my PVE host fails me...)

**If someone can confirm that the syno-way of accelerating HDDs using NVME is not possible inside VDSM that would be great.**

If on top of that you have ideas on how to actually achieve both a virtual syno with great performance and easy fallback to a physical syno..., the only option I see is running a full xpenology wich I deemed too complex for me in my last attempt...

Thanks in advance for any kind of help

Edit: **pretty much same question for USB**: seems passing USB drives to the VM does nothing, probably again nothing in the vDSM official code to support htis use-case right ? So unable to use hyperBackup on USB or similar...

**Comment by mndti** [comment]
You already sort of answered your questions.

Virtual DSM is virtual, it does not have the same features as bare metal DSM.

VDSM being virtual, do you agree that in the end the intention is to pass disks, virtual LAN?

Here what I do is:
NVME - for performance
HDD - for services that do not require performance.

Now if you really want to use 20TB HDD with SSD cache, it's better to go with xpenology bare metal, you can even use VM, but you will suffer in terms of performance, because in the tests I did, they only accept using SATA.

I used https://github.com/AuxXxilium/arc and it works great, it's easy to configure. I just didn't stay, because I needed to use some drivers that unfortunately are not available in DSM.

To pass your discs in VDSM in PVE, as per https://github.com/vdsm/virtual-dsm/issues/674
You should use
boot.img has scsi slot 9
system.img has scsi slot 10
disks scsi slot 11,12,13

**Comment by r0bb10** [comment]
> I was aiming at replacing my physical DS918+ with a VM runing vDSM, happy that if I get a new physical syno NAS in the future I can simply unplug my 20TB drive and replug it into a real syno if they manage to sell me a device with 10G networking at a reasonable price one day...

i stop you right there, you cannot switch from a disk used in a vdsm to a synology dsm, it wont work and you will not see your data in the original syno.

> But seems VDSM does not see it or decides not to use it, or simply does not include the software to make use of nvme read or write cache...

vdsm by architecture does not care about the underlying disk, the correct way is to pass a virtual disk to the vm (vdsm is effectively a vm made my synology) and the performance are tied by the host where the vm (and the vdisk) is hosted.. so if you use proxmox on a raid of whatever disk with cache or not the vdisk passed to vdsm will perform accordingly.

> I understand this makes sens when running vdsm on a syno device as one would pass a volume to vdsm which itself might already be a HDD accelerated with nvme cache, but in my case, if I want vdsm to have good read/write performance I will need to pass it a virtual disk stored manages by my underlying PVE as zfs with caching and stuff... This would also mean that I will not be able to take this physical HDD out of my PVE host and simply plug it into a physical synology if I want to migrate easily in the future (or if my cheap hardware for my PVE host fails me...)

as said, even if you passthrough a whole controller to vdsm with a physical disk attached it will be formatted with a partition table totally different of a baremetal original dsm, and if you move the disk to an original synology data cannot be imported and dsm will ask you to format it.. so no migration, vdsm is not an alternative to dsm, in theory vdsm lives always on top of dsm so it does not need to handle anything hardware and it's not intended to be migrated.

> **If someone can confirm that the syno-way of accelerating HDDs using NVME is not possible inside VDSM that would be great.**

exactly, not possible in the syno-way.. but i have my data disk in vdsm stored on an nvme raid array and it's pretty fast, if you can live with 3500mb/s transfer speed and have the network infrastructure that can handle that speed.

> If on top of that you have ideas on how to actually achieve both a virtual syno with great performance and easy fallback to a physical syno..., the only option I see is running a full xpenology wich I deemed too complex for me in my last attempt...

bare in mind that xpenology is not legal, so not a comparable option.

> Edit: **pretty much same question for USB**: seems passing USB drives to the VM does nothing, probably again nothing in the vDSM official code to support htis use-case right ? So unable to use hyperBackup on USB or similar...

passing vdisk that is on a usb drive can be done and you can use hyperbackup on that.

**Comment by toxic0berliner** [comment]
Ok, thanks a lot everyone for your very clear answers.
I do get that going to VDSM is a one-way road for me, no going back to physical syno easily...
I'm not a fan of the non-legal xpenology but given I feel I gave enough of my money already to syno what bothers me the most is how complex it looks and how "botched together" it all seems, even omitting that it will most probably not let me use nvme caching as well or at least not in a way I'll like.

I will close the question, thanks all, for me it's back to thinking time, I think I really need to bite the bullet and abandon synology, the last thing they hooked me with is hyperbackup, and even that has some limitations, everything ele I already replaced with FOSS and am happy with it...

Will probably go enhance/refine my rsync scripts orchestrated by cronicle-edge and the few duplicati backups I have to replace hyperbackup versioned and monoversion backups...

Sad I couldn't find a one-stop shop solution to run somewhere and tell it "at this frequency you take data from here and copy it up there". backup guys always make things complicated with multiversion and stuff... noone is giving rsync a similar gui to hyperbackup :(
